### DISCUSSION TITLE
Trigger an action upon completion of another action

### DISCUSSION THREAD

**Discussion by cilerler** [community/community#26601] [Actions]
<p>Is there a way to trigger the Action upon completion of another Action in the same repository?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/cilerler">@cilerler</a> ,</p>
<p>Currently, I do not find any build-in option or event to trigger another workflow when a workflow is completed.</p>
<p>I tried the below event configurations in the second workflow:</p>
<pre><code>on:
  check_suite:
    types: [completed]

on:
  check_run:
    types: [completed]
</code></pre>
<p>But if the first workflow has multiple jobs, as long as any one job is completed in the first workflow, the second workflow will be triggered. It will not wait until the completion of the whole first workflow.</p>
<p>As a workaround, you can try to add an additional job at the end of the first workflow, this job is use to trigger the second workflow via the <a href="https://help.github.com/en/actions/reference/events-that-trigger-workflows#external-events-repository_dispatch" rel="nofollow noopener"><strong>repository_dispatch event</strong></a>. In this job, it will <a href="https://developer.github.com/v3/repos/#create-a-repository-dispatch-event" rel="nofollow noopener"><strong>create a repository dispatch event</strong></a>.</p>
<p>A simple demo:</p>
<pre><code># In the first workflow
jobs:
  # some jobs for the first workflow
  job_01:
  job_02:
  job_03:
  . . .
# the job to create repository dispatch event
  job_create_repository_dispatch:
    needs: [job_01, job_02, job_03, ...]
    runs-on: windows-latest
    env:
      GITHUB_PAT: ${{ secrets.GITHUB_PAT }}
    steps:
      - name: Create a repository dispatch event to trigger workflow
        shell: pwsh
        run: |
          $pat = $env:GITHUB_PAT
          $uri = "https://api.github.com/repos/BrightRan/TestClock/dispatches"
          $base64AuthInfo = [Convert]::ToBase64String([Text.Encoding]::ASCII.GetBytes(("{0}:{1}" -f "", $pat)))

          $headers = New-Object "System.Collections.Generic.Dictionary[[String],[String]]"
          $headers.Add("Authorization", ("Basic {0}" -f $base64AuthInfo))
          $headers.Add("Accept", "application/vnd.github.everest-preview+json")
          $headers.Add("Content-Type", "application/json")

          $body = "{
            `"event_type`": `"test-repository-dispatch`",
            `"client_payload`": {
              `"unit`": false,
              `"integration`": true
            }
          }"

          $response = Invoke-RestMethod -Uri $uri -Headers $headers -Body $body -Method POST
          $response | ConvertTo-Json

# In the sexond workflow
on:
  repository_dispatch:
</code></pre>
<p>In this situation, on the first workflow, when all the previous jobs ( <strong>job_01</strong> , <strong>job_02</strong> , <strong>job_03</strong> , <strong>…</strong> ) complete successfully, the job  <strong>job_create_repository_dispatch </strong> will start and execute the GitHub REST API to create a repository dispatch event, then the second workflow is triggered to run.</p>
<p><strong>Note:</strong> About the <strong>GITHUB_PAT</strong> used in the workflow, you need to create a personal access token and set it as a secret in your repository, instead of directly using the <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/authenticating-with-the-github_token" rel="nofollow noopener"><strong>GITHUB_TOKEN</strong></a> automatically created by GitHub.</p>

**Comment by cilerler** [comment]
<p><a class="mention" href="/brightran">@brightran</a> I appreciate it.  Your answer is very detailed and on the spot.  Thank you!</p>
<p>The only missing part in your example is how to read the payload.</p>
<p>Let’s consider that the payload will be </p>
<pre><code>$body = "{
            `"event_type`": `"test-repository-dispatch`",
            `"client_payload`": {
              `"completedTask`": "library1"
            }
          }"
</code></pre>
<p>Now, how will I be able to read it from the second workflow? e.g.</p>
<pre><code>name: library2

on:
  repository_dispatch:

jobs:
build:
if: github.event_name == 'push' &amp;&amp; contains(toJson(github.?.?.?.completedTask), 'library1')
runs-on: windows-latest
...
</code></pre>
<p>what it should be should instead of <code>?.?.?</code></p>
<p>Thanks in advance.</p>

**Comment by BrightRan** [comment]
<p>Yes, in the second workflow which is triggered via the repository_dispatch event you can get the  <strong>client_payload</strong> from the <a href="https://help.github.com/en/actions/reference/contexts-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener"><strong>github context</strong></a> of the second workflow. You can use the property  <strong>github.event.client_payload</strong> to access the object  <strong>client_payload</strong>. And you also can use the similar syntaxes to access the children objects of <strong>client_payload</strong> , such as  <strong>github.event.client_payload.integration</strong> in my example.</p>
<p>The property  <strong>github.event.action</strong> is the  <strong>event_type</strong> you set when creating the repository dispatch event in the first workflow, such as  <strong>test-repository-dispatch</strong> in my example.</p>
<p>You can reference the <a href="https://help.github.com/en/actions/reference/contexts-and-expression-syntax-for-github-actions#example-printing-context-information-to-the-log-file" rel="nofollow noopener"><strong>example printing context information to the log file</strong></a> to print the github context in the second workflow.</p>
<p>For example:</p>
<pre><code>on:
  repository_dispatch:

jobs:
  workflow_run_info:
    name: Information about the workflow run
    runs-on: ubuntu-latest
    env:
      GITHUB_CONTEXT: ${{ toJson(github) }}
    steps:
      - name: Show Information
        run: |
          echo "This workflow run is triggered by ${{ github.event_name }} ."
          echo " *****************************************************************************"
          echo "$GITHUB_CONTEXT"
          echo " *****************************************************************************"
</code></pre>
<p>Logs:</p>
<p><img src="https://user-images.githubusercontent.com/46307996/181098637-b62046be-fb55-4c4d-ace1-fc021286dd05.png" alt="github_context.png" data-base62-sha1="gfjtPZw28C8TnLJseQf76waxEj6" width="455" height="420"></p>

**Comment by cilerler** [comment]
<p>I tried to implement it, but it doesn’t fit what I would like to do.  The way you did it is merely signaling the next action. The issue is, I have no control knowledge of which action should be triggered.</p>
<p><strong>Real-life scenario;</strong></p>
<p>I have an action that creates a based library package that holds all the models.  Upon the package update, I would like to redeploy all the other packages (which are created by other actions) that use the package. </p>
<p>It could be done in a single action if there was a way to filter the triggering path in the steps as we do on top of the action. However, right now, there is no solution for me to have a robust way of deploying all the packages and not updating the ones that don’t need any update.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/cilerler">@cilerler</a> ,</p>
<p>According to your request, maybe you can try the below steps:</p>
<ol>
<li>
<p>Every time when the based library package updated, in the workflow for the based library package, when configuring the API to create the repository dispatch event, you can custom an object in the <strong>client_payload</strong> to specify which workflows will be run.<br>
For example, custom the object <strong>run_workflows</strong> :</p>
<p>{<br>
“event_type”: “test-repository-dispatch”,<br>
“client_payload”: {<br>
“unit”: false,<br>
“integration”: true,<br>
“run_workflows”: [<br>
{“name”: “workflow1”},<br>
{“name”: “workflow2”},<br>
{“name”: “workflow3”},<br>
. . .<br>
{“name”: “workflowN”}<br>
]<br>
}<br>
}</p>
</li>
<li>
<p>In the workflows for other packages, set they are triggered <strong>on.repository_dispatch</strong> event. Add the below if conditional on all the deploy jobs of the workflows.</p>
<p>if: contains(github.event.client_payload.run_workflows.*.name, github.workflow)</p>
</li>
</ol>
<p>This if conditional will check if the currently running workflow is contained in <strong>run_workflows</strong> you set.<br>
<strong>github.event.client_payload.run_workflows.*.name</strong> returns all names of the workflows as an array.<br>
<strong>github.workflow</strong> returns the name of the currently running workflow.</p>
<p>When the <a href="https://help.github.com/en/actions/reference/contexts-and-expression-syntax-for-github-actions#contains" rel="nofollow noopener"><strong>contains()</strong></a> function retuns ’ <strong>true</strong>’, the deploy job in current workflow will be executed, if ’ <strong>false</strong>’, the deploy job will be skipped.</p>

**Comment by cilerler** [comment]
<p>I hear you, and I appreciate all the details you provided.  Unfortunately, your logic still relies on me placing consumer workflows into the base workflow. Where it is not a good practice, no base library should know who will consume it.  Even further maintenance of nested triggers would be a nightmare. For that reason, I believe this should be a feature request. </p>
<p>My 2 cents say this should be a part of the internal dynamics of the GitHub Actions, where workflows should be able to tap into other workflows and should be triggered upon successful completion.  The only major issue that I can see would be a circular-reference, where it could be a deal-breaker. </p>
<p>e.g.</p>
<pre><code>name: library2

on:
  completion:
    success:
    - library1
  push:
    branches:
    - master
    paths:
    - '.github/workflows/library2.yml'
    - 'src/library2/**'
</code></pre>
<p>Unfortunately, I don’t see any clear path to deploy library packages via GitHub Actions the way it lets us at this point. <a href="https://github.com/cilerler/ruya/tree/master/src" rel="nofollow noopener">Here</a> is a real-life scenario, each project should generate its own packages, and consider nested triggers can occur.</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/cilerler">@cilerler</a> ,</p>
<p>Sorry my suggestions did not help you solve the problem.</p>
<p>You request is quite similar to the trigger option " <strong>Build completion</strong>" on <strong>Azure Pipelines</strong>. However, currently GitHub Actions does not have a build-in option like this, and it seems that the only way we can use to trigger a workflow from another workflow is the <a href="https://help.github.com/en/actions/reference/events-that-trigger-workflows#external-events-repository_dispatch" rel="nofollow noopener"><strong>repository_dispatch event</strong></a>.</p>
<p><img src="https://user-images.githubusercontent.com/46307996/181098640-41696b03-71eb-419d-9235-ef5396872bd5.png" alt="azure_pipelines.png" data-base62-sha1="jZbMko77CoFZbvIUcdMosgfE6LM" width="265" height="309"></p>
<p>If your projects really need this feature, I recommend you directly report a feature request <a href="https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions" rel="nofollow noopener">here</a>. That will allow you to directly interact with the appropriate engineering team, and make it more convenient for the engineering team to collect and categorize your suggestions.</p>

**Comment by cilerler** [comment]
<p>Hi <a class="mention" href="/brightran">@brightran</a> </p>
<p>Nice catch! That is precisely what I’m looking for to move from Azure Pipelines to GitHub Actions.</p>
<p>As I stated earlier, the repository_dispatch event is a <code>push</code> based concept, where I’m looking for <code>pull</code>.</p>
<p>I appreciate all the help you provided. Thank you!</p>

**Comment by Molkree** [comment]
<p><a class="mention" href="/cilerler">@cilerler</a>, if you don’t want to put consumer workflows into your base workflow, can’t you do the reverse? Send repository_dispatch event when base workflow finishes, run your consumer workflows only on that repository_workflow type.<br>
In your base workflow:</p>
<pre><code class="lang-yaml">- name: Send repository_dispatch
  env:
    PAT: ${{ secrets.ACTIONS_PAT }}
  run: |
    curl \
      -X POST \
      -H "Authorization: token $PAT" \
      -H "Accept: application/vnd.github.v3+json" \
      https://api.github.com/repos/{owner}/{repo}/dispatches \
      -d '{"event_type":"event_type"}'
</code></pre>
<p>and in workflows that should run after base:</p>
<pre><code class="lang-yaml">on:
  repository_dispatch:
    types: [event_type]
</code></pre>
<p>PAT should be with <code>repo</code> scope. <code>event_type</code> can be whatever you decide (<a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#repository_dispatch" rel="noopener nofollow ugc">docs</a>).</p>
<p>See it working <a href="https://github.com/Molkree/github_actions" rel="noopener nofollow ugc">here</a>.</p>
<p>P.S. After finishing my message I realized this is over 1 year old (2020-&gt;2021 lol), but it’s actually the first result for me when searching “trigger github action from another action”.</p>
<p>P.P.S. Or you can use new <code>workflow_dispatch</code> event. It will be less flexible in the way that it will run even if your base workflow fails, but you will be able to check if it failed inside consumer workflow.</p>
<pre><code class="lang-yaml">on:
  workflow_run:
    workflows: ["Build"]
    types: [completed]

jobs:
  on-success:
    runs-on: ubuntu-latest
    if: ${{ github.event.workflow_run.conclusion == 'success' }}
    steps:
      ...
  on-failure:
    runs-on: ubuntu-latest
    if: ${{ github.event.workflow_run.conclusion == 'failure' }}
    steps:
      ...
</code></pre>
<p><a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_run" rel="noopener nofollow ugc">Docs</a></p>

**Comment by cilerler** [comment]
Unfortunately, I couldn't see your recent activity on the issue, so I'm unsure if you're still available to help.

Regardless, I wanted to thank you for your previous response. However, I wanted to let you know that the solution you proposed won't work as I still need to provide the target repository name in the source repository to trigger the event.

If you have any further suggestions, I'd love to hear them. Otherwise, I'll keep looking for a solution and update the issue as necessary.

**Comment by amuthan-sakthivel** [comment]
https://github.com/peter-evans/repository-dispatch you can simply use this action at the end of the first workflow to trigger the second workflow

youtube tutorial - https://www.youtube.com/watch?v=fOG1C_6yjvQ&list=PL9ok7C7Yn9A-6uidd3RXZPf5EfhxkPXa_&index=11

**Comment by cilerler** [comment]
Thank you, @amuthan-sakthivel! I appreciate your suggestion, and I'm happy to share that I've already resolved the issue using the same library you mentioned.
